crates/symbolicator/src/services/symbolication.rs

```diff
@@ -1768,6 +1859,113 @@ fn load_cfi_for_processor(cfi: BTreeMap<DebugId, PathBuf>) -> BTreeMap<DebugId,
         .collect()
 }
 
+struct TempSymbolProvider {
+    files: BTreeMap<DebugId, SymbolFile>,
+    missing_ids: RefCell<BTreeSet<DebugId>>,
+}
+
+impl TempSymbolProvider {
+    /// Load the CFI information from the cache.
+    ///
+    /// This reads the CFI caches from disk and returns them in a format suitable for the
+    /// breakpad processor to stackwalk.
+    pub fn new<'a, M: Iterator<Item = (&'a DebugId, &'a PathBuf)>>(modules: M) -> Self {
+        // TODO(ja): Make TempSymbolProvider the thing serialized to procspawn (prepares for moving in-process)
+        Self {
+            files: modules
+                .filter_map(|(id, path)| Some((*id, Self::load(path)?)))
+                .collect(),
+
+            missing_ids: RefCell::new(BTreeSet::new()),
+        }
+    }
+
+    fn load(cfi_path: &Path) -> Option<SymbolFile> {
+        let bytes = ByteView::open(cfi_path)
+            .map_err(|err| {
+                let stderr: &dyn std::error::Error = &err;
+                tracing::error!(stderr, "Error while reading cficache");
+            })
+            .ok()?;
+
+        let cfi_cache = CfiCache::from_bytes(bytes)
+            // This mostly never happens since we already checked the files
+            // after downloading and they would have been tagged with
+            // CacheStatus::Malformed.
+            .map_err(|err| {
+                let stderr: &dyn std::error::Error = &err;
+                tracing::error!(stderr, "Error while loading cficache");
+            })
+            .ok()?;
+
+        SymbolFile::from_bytes(cfi_cache.as_slice())
+            .map_err(|err| {
+                let stderr: &dyn std::error::Error = &err;
+                tracing::error!(stderr, "Error while procecssing cficache");
+            })
+            .ok()
+    }
+
+    fn missing_ids(&self) -> BTreeSet<DebugId> {
+        std::mem::take(&mut *self.missing_ids.borrow_mut())
+    }
+}
+
+impl SymbolProvider for TempSymbolProvider {
```

Comment:
We are not using `stats` ourselves, so I wonder if that trait method is (or should be) optional?

Other than that, we should also petition @gankra to make the trait methods take `&mut self`, so we can actually lazy load stuff without relying on `RefCell` and similar.

Reply:
*macos beachball forms as I try to remember why everything in breakpad-symbols was written this way*

I *think* Ted just had aspirations of allowing this stuff to be multithreaded (but I am totally fine with never going down that path). I'm not sure if there's some situation where we are actually leveraging this flexibility externally. I know *internally* it's a bit of a mess and have vague memories of running into weird ownership stuff, but 50/50 on that just being *induced* by the RefCells.

In my mind making everything &mut should *conceptually* be totally fine, but I wouldn't be surprised if actually trying to do it would have some nasty roadblocks.

As for stats: minidump-processor's JSON output sources some data from it, so I would *prefer* that you keep implementing those parts? That said a lot of the stats infra was me concepting out some stuff I never got around to implementing.